Fix painting of RenderListBox in vertical writing mode

pxlcoder opened the pull request:
#### c7a54f98d7217bc27161357670a5b7549a231959
<pre>
Fix painting of RenderListBox in vertical writing mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=261802">https://bugs.webkit.org/show_bug.cgi?id=261802</a>
rdar://115766419

Reviewed by Tim Nguyen.

Ensure option items are painted following the block direction. This patch
also renames variables refer to logical values where necessary.

Subsequent patches will add support for horizontal scrolling and keyboard
navigation.

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-multiple-options-visual-order-expected.txt: Added.

The newly added tests is a failures for now, since `&lt;select&gt;` is still forced
to `writing-mode: horizontal-tb`. Once `RenderListBox` fully supports vertical
writing mode, and is enabled under the setting, these tests will pass.

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-multiple-options-visual-order.html: Added.
* LayoutTests/platform/ios/TestExpectations:

`RenderListBox` is not used on iOS.

* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::updateFromElement):
(WebCore::RenderListBox::layout):
(WebCore::RenderListBox::computeIntrinsicLogicalWidths const):
(WebCore::RenderListBox::computePreferredLogicalWidths):
(WebCore::RenderListBox::numVisibleItems const):
(WebCore::RenderListBox::listLogicalHeight const):
(WebCore::RenderListBox::computeLogicalHeight const):
(WebCore::RenderListBox::itemBoundingBoxRect const):
(WebCore::RenderListBox::paintItem):
(WebCore::itemOffsetForAlignment):
(WebCore::RenderListBox::paintItemForeground):
(WebCore::RenderListBox::listIndexAtOffset const):
(WebCore::RenderListBox::panScroll):
(WebCore::RenderListBox::listIndexIsVisible):
(WebCore::RenderListBox::maximumNumberOfItemsThatFitInPaddingAfterArea const):
(WebCore::RenderListBox::numberOfVisibleItemsInPaddingBefore const):
(WebCore::RenderListBox::numberOfVisibleItemsInPaddingAfter const):
(WebCore::RenderListBox::computeFirstIndexesVisibleInPaddingBeforeAfterAreas):
(WebCore::RenderListBox::scrollTo):
(WebCore::RenderListBox::itemLogicalHeight const):
(WebCore::RenderListBox::scrollHeight const):
(WebCore::RenderListBox::scrollTop const):
(WebCore::RenderListBox::setScrollTop):
(WebCore::RenderListBox::listHeight const): Deleted.
(WebCore::RenderListBox::maximumNumberOfItemsThatFitInPaddingBottomArea const): Deleted.
(WebCore::RenderListBox::numberOfVisibleItemsInPaddingTop const): Deleted.
(WebCore::RenderListBox::numberOfVisibleItemsInPaddingBottom const): Deleted.
(WebCore::RenderListBox::computeFirstIndexesVisibleInPaddingTopBottomAreas): Deleted.
(WebCore::RenderListBox::itemHeight const): Deleted.
* Source/WebCore/rendering/RenderListBox.h:

Canonical link: <a href="https://commits.webkit.org/268594@main">https://commits.webkit.org/268594@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a00788ecd65f2691cb3db6c1443bd3840596e417

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20042 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20481 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21099 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21941 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18710 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20272 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20639 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20199 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20190 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17418 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22791 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17360 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18215 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24472 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18437 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18391 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22462 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18984 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16104 "5 flakes 1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18179 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18051 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4827 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22524 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18810 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -1464,6 +1464,10 @@ webkit.org/b/133057 fast/table/border-collapsing/collapsed-borders-adjoining-sec
 
+# <select> element vertical writing mode support is not yet enabled
+fast/forms/vertical-writing-mode/listbox-vertical-lr-option-order.html [ Failure ]
+fast/forms/vertical-writing-mode/listbox-vertical-rl-option-order.html [ Failure ]
```
These new tests could be WPT.

nt1m replied:
See:

https://github.com/web-platform-tests/wpt/blob/master/css/css-writing-modes/forms/select-size-scrolling-and-sizing.optional.html
https://github.com/web-platform-tests/wpt/blob/master/css/css-writing-modes/forms/select-multiple-scrolling.optional.html

pxlcoder replied:
Rewrote as a WPT.

pxlcoder committed:
Fix painting of RenderListBox in vertical writing mode
https://bugs.webkit.org/show_bug.cgi?id=261802
rdar://115766419

Reviewed by Tim Nguyen.

Ensure option items are painted following the block direction. This patch
also renames variables refer to logical values where necessary.

Subsequent patches will add support for horizontal scrolling and keyboard
navigation.

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-multiple-options-visual-order-expected.txt: Added.

The newly added tests is a failures for now, since `<select>` is still forced
to `writing-mode: horizontal-tb`. Once `RenderListBox` fully supports vertical
writing mode, and is enabled under the setting, these tests will pass.

* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/select-multiple-options-visual-order.html: Added.
* LayoutTests/platform/ios/TestExpectations:

`RenderListBox` is not used on iOS.

* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::updateFromElement):
(WebCore::RenderListBox::layout):
(WebCore::RenderListBox::computeIntrinsicLogicalWidths const):
(WebCore::RenderListBox::computePreferredLogicalWidths):
(WebCore::RenderListBox::numVisibleItems const):
(WebCore::RenderListBox::listLogicalHeight const):
(WebCore::RenderListBox::computeLogicalHeight const):
(WebCore::RenderListBox::itemBoundingBoxRect const):
(WebCore::RenderListBox::paintItem):
(WebCore::itemOffsetForAlignment):
(WebCore::RenderListBox::paintItemForeground):
(WebCore::RenderListBox::listIndexAtOffset const):
(WebCore::RenderListBox::panScroll):
(WebCore::RenderListBox::listIndexIsVisible):
(WebCore::RenderListBox::maximumNumberOfItemsThatFitInPaddingAfterArea const):
(WebCore::RenderListBox::numberOfVisibleItemsInPaddingBefore const):
(WebCore::RenderListBox::numberOfVisibleItemsInPaddingAfter const):
(WebCore::RenderListBox::computeFirstIndexesVisibleInPaddingBeforeAfterAreas):
(WebCore::RenderListBox::scrollTo):
(WebCore::RenderListBox::itemLogicalHeight const):
(WebCore::RenderListBox::scrollHeight const):
(WebCore::RenderListBox::scrollTop const):
(WebCore::RenderListBox::setScrollTop):
(WebCore::RenderListBox::listHeight const): Deleted.
(WebCore::RenderListBox::maximumNumberOfItemsThatFitInPaddingBottomArea const): Deleted.
(WebCore::RenderListBox::numberOfVisibleItemsInPaddingTop const): Deleted.
(WebCore::RenderListBox::numberOfVisibleItemsInPaddingBottom const): Deleted.
(WebCore::RenderListBox::computeFirstIndexesVisibleInPaddingTopBottomAreas): Deleted.
(WebCore::RenderListBox::itemHeight const): Deleted.
* Source/WebCore/rendering/RenderListBox.h:

Canonical link: https://commits.webkit.org/268594@main
